core/src/autogluon/core/models/abstract/abstract_model.py

```diff
@@ -520,6 +520,84 @@ def fit(self, **kwargs):
             logger.warning(f'\tWarning: Model has no time left to train, skipping model... (Time Left = {round(kwargs["time_limit"], 1)}s)')
             raise TimeLimitExceeded
 
+    def fit_with_prune(self, X, y, X_val, y_val, max_num_fit=3, prune_threshold=0., stop_threshold=3, num_shuffle_sets=3, **kwargs):
+        """
+        Functionally identical to `fit` method, but repeats feature importance based pruning until
+        validation set performance degrades or `max_num_fit` iterations have passed.
+
+        Parameters
+        ----------
+        max_num_fit : int, default = 3
+            Maximum number of time feature selection and fitting are performed
+        prune_threshold : float, default = 0.
+            Feature importance threshold that features must meet in order to not be dropped
+        stop_threshold : int, default = 3
+            Early stopping will stop refitting model if score does not improve for this amount of iterations
+        num_shuffle_sets : int, default = 3
+            Number of shuffles to evaluate when computing permutation feature importance
+
+        TODO
+        1. Respect time limit
+        2. Consider returning list of models
+
+        FIXME
+        1. Does not work with S3 paths (no problem locally)
+        """
+        if X_val is None or y_val is None:
+            raise AssertionError(f'fit_with_prune must be called with validation data.')
+
+        best_score = None
+        best_index = None
+        fitted_copies_info = []
+        current_features = None
+        try:
+            for index in range(max_num_fit):
+                # Fit a deepcopied model on current_features
+                curr_model_path = os.path.join(self.path, f'prune_{index}', '')
+                self_copy = self.convert_to_template()
+                self_copy.features = current_features
+                self_copy.path = curr_model_path
+                self_copy.fit(X=X, y=y, X_val=X_val, y_val=y_val, **kwargs)
+                self_copy.save()
+
+                score = self_copy.score(X=X_val, y=y_val)
```

Comment:
To enable for bagging:

If X_val=None, y_val=None -> is_oof=True, then call self_copy.score_with_oof(y=y)

Reply:
Add TODO: if self.weight_evaluation==True pass sample_weight param for scoring